src/block/Composter.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\block\utils\BlockDataSerializer;
+use pocketmine\block\utils\ComposterUtils;
+use pocketmine\entity\projectile\Projectile;
+use pocketmine\item\Fertilizer;
+use pocketmine\item\Item;
+use pocketmine\item\ItemIdentifier;
+use pocketmine\item\ItemIds;
+use pocketmine\math\AxisAlignedBB;
+use pocketmine\math\Facing;
+use pocketmine\math\Vector3;
+use pocketmine\network\mcpe\protocol\SpawnParticleEffectPacket;
+use pocketmine\network\mcpe\protocol\types\DimensionIds;
+use pocketmine\player\Player;
+use pocketmine\world\sound\ComposterEmptySound;
+use pocketmine\world\sound\ComposterFillSound;
+use pocketmine\world\sound\ComposterFillSuccessSound;
+use pocketmine\world\sound\ComposterReadySound;
+use function abs;
+use function mt_rand;
+
+class Composter extends Transparent{
+	public const CROP_GROWTH_EMITTER_PARTICLE = "minecraft:crop_growth_emitter";
+
+	protected int $composter_fill_level = 0;
+
+	protected function writeStateToMeta() : int{
+		return $this->composter_fill_level;
+	}
+
+	public function writeStateToItemMeta() : int{
+		return $this->composter_fill_level;
+	}
+
+	public function readStateFromData(int $id, int $stateMeta) : void{
+		$this->composter_fill_level = BlockDataSerializer::readBoundedInt("composter_fill_level", $stateMeta, 0, 8);
+	}
+
+	public function getStateBitmask() : int{
+		return 0b1111;
+	}
+
+	protected function recalculateCollisionBoxes() : array{
+		$boxes = [$this->getSideCollisionBox(Facing::DOWN)];
+		foreach (Facing::HORIZONTAL as $side) {
+			$boxes[] = $this->getSideCollisionBox($side);
+		}
+		return $boxes;
+	}
+
+	protected function getSideCollisionBox(int $face = Facing::NORTH) : AxisAlignedBB{
+		$empty = abs(15 - 2 * $this->composter_fill_level) - (int) ($this->composter_fill_level === 0);
```

Comment:
I have no idea what's going on here.

Reply:
basically calculate each side collision boxes, so i was thinking of making 4 walls each side and one layer at the bottom which rise up when the composter get compost

Reply:
I think changing the `abs` to `max(0, …)` would be a better way

Reply:
> basically it calculates each side collision boxes, so i was thinking of making 4 walls each side and one layer at the bottom which will rise up when the composter get composted

You missed my point entirely. The code should be self explanatory and it's not because you tried too hard to be clever.